src/theory/arith/nl/piand_solver.cpp

```diff
@@ -71,13 +71,145 @@ void PIAndSolver::initLastCall(const std::vector<Node>& assertions,
                  << std::endl;
 }
 
-void PIAndSolver::checkInitialRefine() {}
+void PIAndSolver::checkInitialRefine()
+{
+  Trace("piand-check") << "PIAndSolver::checkInitialRefine" << std::endl;
+  NodeManager* nm = nodeManager();
+  for (const std::pair<const Node, std::vector<Node> >& is : d_piands)
+  {
+    // the reference bitwidth
+    Node k = is.first;
+    for (const Node& i : is.second)
+    {
+      Node x = i[1];
+      Node y = i[2];
+      if (d_initRefine.find(i) != d_initRefine.end())
+      {
+        // already sent initial axioms for i in this user context
+        continue;
+      }
+      d_initRefine.insert(i);
+      Node twok = nm->mkNode(Kind::POW2, k);
+      Node arg0Mod = nm->mkNode(Kind::INTS_MODULUS, x, twok);
+      Node arg1Mod = nm->mkNode(Kind::INTS_MODULUS, y, twok);
+      Node arg0Mod2 = nm->mkNode(Kind::INTS_MODULUS, x, d_two);
+      Node arg1Mod2 = nm->mkNode(Kind::INTS_MODULUS, y, d_two);
+      Node plus = nm->mkNode(Kind::ADD, x, y);
+      Node twok_minus_one = nm->mkNode(Kind::SUB, twok, d_one);
+      Node k_gt_0 = nm->mkNode(Kind::GT, k, d_zero);
+      Node x_geq_zero = nm->mkNode(Kind::GEQ, x, d_zero);
+      Node x_lt_pow2 = nm->mkNode(Kind::LT, x, twok);
+      Node x_range = nm->mkNode(Kind::AND, x_geq_zero, x_lt_pow2);
+      Node y_geq_zero = nm->mkNode(Kind::GEQ, y, d_zero);
+      Node y_lt_pow2 = nm->mkNode(Kind::LT, y, twok);
+      Node y_range = nm->mkNode(Kind::AND, y_geq_zero, y_lt_pow2);
+      // initial refinement lemmas
+      std::vector<Node> conj;
+      Assert(x <= y);
```

Comment:
explain, and also, i am not sure this really holds right now without the rewriter.